actix-utils/src/framed.rs

```diff
@@ -62,60 +59,61 @@ pub enum Message<T> {
     Close,
 }
 
-/// FramedTransport - is a future that reads frames from Framed object
-/// and pass then to the service.
+/// Dispatcher is a future that reads frames from Framed object
+/// and passes them to the service.
 #[pin_project::pin_project]
-pub struct Dispatcher<S, T, U>
+pub struct Dispatcher<S, T, U, I>
 where
-    S: Service<Request = Request<U>, Response = Response<U>>,
+    S: Service<Request = <U as Decoder>::Item, Response = I>,
     S::Error: 'static,
     S::Future: 'static,
     T: AsyncRead + AsyncWrite,
-    U: Encoder + Decoder,
-    <U as Encoder>::Item: 'static,
-    <U as Encoder>::Error: std::fmt::Debug,
+    U: Encoder<I> + Decoder,
+    // I: 'static,
+    <U as Encoder<I>>::Error: std::fmt::Debug,
 {
     service: S,
-    state: State<S, U>,
+    state: State<S, U, I>,
     #[pin]
     framed: Framed<T, U>,
-    rx: mpsc::Receiver<Result<Message<<U as Encoder>::Item>, S::Error>>,
-    tx: mpsc::Sender<Result<Message<<U as Encoder>::Item>, S::Error>>,
+    rx: mpsc::Receiver<Result<Message<I>, S::Error>>,
+    tx: mpsc::Sender<Result<Message<I>, S::Error>>,
 }
 
-enum State<S: Service, U: Encoder + Decoder> {
+enum State<S: Service, U: Encoder<I> + Decoder, I> {
     Processing,
-    Error(DispatcherError<S::Error, U>),
-    FramedError(DispatcherError<S::Error, U>),
+    Error(DispatcherError<S::Error, U, I>),
+    FramedError(DispatcherError<S::Error, U, I>),
     FlushAndStop,
     Stopping,
 }
 
-impl<S: Service, U: Encoder + Decoder> State<S, U> {
-    fn take_error(&mut self) -> DispatcherError<S::Error, U> {
+impl<S: Service, U: Encoder<I> + Decoder, I> State<S, U, I> {
+    fn take_error(&mut self) -> DispatcherError<S::Error, U, I> {
         match mem::replace(self, State::Processing) {
             State::Error(err) => err,
             _ => panic!(),
         }
     }
 
-    fn take_framed_error(&mut self) -> DispatcherError<S::Error, U> {
+    fn take_framed_error(&mut self) -> DispatcherError<S::Error, U, I> {
         match mem::replace(self, State::Processing) {
             State::FramedError(err) => err,
             _ => panic!(),
         }
     }
 }
 
-impl<S, T, U> Dispatcher<S, T, U>
+impl<S, T, U, I> Dispatcher<S, T, U, I>
 where
-    S: Service<Request = Request<U>, Response = Response<U>>,
+    S: Service<Request = <U as Decoder>::Item, Response = I>,
     S::Error: 'static,
     S::Future: 'static,
     T: AsyncRead + AsyncWrite,
-    U: Decoder + Encoder,
-    <U as Encoder>::Item: 'static,
-    <U as Encoder>::Error: std::fmt::Debug,
+    U: Decoder + Encoder<I>,
+    // I: 'static,
```

Comment:
This could be uncommented.

```suggestion
    I: 'static,
```